File: sdk/ts/src/skills.ts
// Skills API - §28 §3.6 packaging spec
// 注意: /sisoul/skill/* 路径已含 /sisoul 前缀, 走 absolute=true.

import type { SisoulClient } from "./client.js";
import type {
  SkillBorrowRequest,
  SkillBorrowResponse,
  SkillCreateRequest,
  SkillCreateResponse,
  SkillLendRequest,
  SkillLendResponse,
  SkillListResponse,
  SkillSessionsResponse,
  SkillItem,
  SkillSessionItem,
} from "./types.js";

export interface EndSessionResponse {
  session_id: string;
  status: string;
  destroy_reason?: string;
  destroyed_at?: number;
  ledger_entry_id?: string;
}

export class SkillsAPI {
  constructor(private readonly client: SisoulClient) {}

  async list(): Promise<SkillListResponse> {
    return this.client.get<SkillListResponse>("/sisoul/skill/list", { absolute: true });
  }

  async owned(): Promise<SkillItem[]> {
    return (await this.list()).owned;
  }

  async available(): Promise<SkillItem[]> {
    return (await this.list()).available_to_borrow;
  }

  async create(req: SkillCreateRequest): Promise<SkillCreateResponse> {
    if (!req.name) throw new Error("skills.create: name required");
    if (!req.system_prompt) throw new Error("skills.create: system_prompt required");
    return this.client.post<SkillCreateResponse>("/sisoul/skill/create", req, {
      absolute: true,
    });
  }

  async lend(req: SkillLendRequest): Promise<SkillLendResponse> {
    if (!req.skill_id) throw new Error("skills.lend: skill_id required");
    if (!req.permissions?.mode) throw new Error("skills.lend: permissions.mode required");
    return this.client.post<SkillLendResponse>("/sisoul/skill/lend", req, {
      absolute: true,
    });
  }

  async borrow(req: SkillBorrowRequest): Promise<SkillBorrowResponse> {
    if (!req.owner_did) throw new Error("skills.borrow: owner_did required");
    if (!req.qualified_name) throw new Error("skills.borrow: qualified_name required");
    return this.client.post<SkillBorrowResponse>("/sisoul/skill/borrow", req, {
      absolute: true,
    });
  }

  async sessions(): Promise<SkillSessionItem[]> {
    const r = await this.client.get<SkillSessionsResponse>("/sisoul/skill/sessions", {
      absolute: true,
    });
    return r.sessions ?? [];
  }

  async activeSessions(): Promise<SkillSessionItem[]> {
    return (await this.sessions()).filter((s) => s.status === "active");
  }

  async endSession(session_id: string, reason?: string): Promise<EndSessionResponse> {
    if (!session_id) throw new Error("skills.endSession: session_id required");
    return this.client.post<EndSessionResponse>(
      "/sisoul/skill/end-session",
      { session_id, reason },
      { absolute: true }
    );
  }
}

File: sdk/rust/src/skills.rs
//! Skills API — §28 §3.6 packaging spec.
//!
//! 注意: /sisoul/skill/* 路径已含 /sisoul 前缀, 用 absolute=true.

use crate::client::SisoulClient;
use crate::errors::{Result, SisoulError};
use crate::types::{
    EndSessionRequest, EndSessionResponse, SkillBorrowRequest, SkillBorrowResponse,
    SkillCreateRequest, SkillCreateResponse, SkillItem, SkillLendRequest, SkillLendResponse,
    SkillListResponse, SkillSessionItem, SkillSessionStatus, SkillSessionsResponse,
};

pub struct SkillsAPI<'a> {
    c: &'a SisoulClient,
}

impl<'a> SkillsAPI<'a> {
    pub fn new(c: &'a SisoulClient) -> Self {
        Self { c }
    }

    pub fn list(&self) -> Result<SkillListResponse> {
        self.c.get_json("/sisoul/skill/list", &[], true)
    }

    pub fn owned(&self) -> Result<Vec<SkillItem>> {
        Ok(self.list()?.owned)
    }

    pub fn available(&self) -> Result<Vec<SkillItem>> {
        Ok(self.list()?.available_to_borrow)
    }

    pub fn create(&self, req: &SkillCreateRequest) -> Result<SkillCreateResponse> {
        if req.name.is_empty() {
            return Err(SisoulError::InvalidArgument("skills.create: name required".into()));
        }
        if req.system_prompt.is_empty() {
            return Err(SisoulError::InvalidArgument(
                "skills.create: system_prompt required".into(),
            ));
        }
        self.c.post_json("/sisoul/skill/create", req, true)
    }

    pub fn lend(&self, req: &SkillLendRequest) -> Result<SkillLendResponse> {
        if req.skill_id.is_empty() {
            return Err(SisoulError::InvalidArgument("skills.lend: skill_id required".into()));
        }
        self.c.post_json("/sisoul/skill/lend", req, true)
    }

    pub fn borrow(&self, req: &SkillBorrowRequest) -> Result<SkillBorrowResponse> {
        if req.owner_did.is_empty() {
            return Err(SisoulError::InvalidArgument("skills.borrow: owner_did required".into()));
        }
        if req.qualified_name.is_empty() {
            return Err(SisoulError::InvalidArgument(
                "skills.borrow: qualified_name required".into(),
            ));
        }
        self.c.post_json("/sisoul/skill/borrow", req, true)
    }

    pub fn sessions(&self) -> Result<Vec<SkillSessionItem>> {
        let r: SkillSessionsResponse = self.c.get_json("/sisoul/skill/sessions", &[], true)?;
        Ok(r.sessions)
    }

    pub fn active_sessions(&self) -> Result<Vec<SkillSessionItem>> {
        Ok(self
            .sessions()?
            .into_iter()
            .filter(|s| s.status == SkillSessionStatus::Active)
            .collect())
    }

    pub fn end_session(&self, session_id: &str, reason: Option<&str>) -> Result<EndSessionResponse> {
        if session_id.is_empty() {
            return Err(SisoulError::InvalidArgument(
                "skills.end_session: session_id required".into(),
            ));
        }
        let body = EndSessionRequest {
            session_id: session_id.to_string(),
            reason: reason.map(String::from),
        };
        self.c.post_json("/sisoul/skill/end-session", &body, true)
    }
}
